src/v/iceberg/metadata_query.h

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+#include "base/seastarx.h"
+#include "iceberg/manifest_io.h"
+#include "iceberg/table_metadata.h"
+
+#include <seastar/core/future.hh>
+
+namespace iceberg {
+
+/**
+ * Enum class allowing changing an expected type returned from query
+ */
+enum class result_type { snapshot, manifest_file, manifest };
+
+template<typename ElemT>
+using matcher = absl::FunctionRef<bool(const ElemT&)>;
+/**
+ * Metadata query representing data structure. Query result_type template
+ * parameter controls what information should be returned after query is
+ * executed. The matchers are predicates returning `true` if an entry should be
+ * included in the result. Empty matcher matches all entities. Matchers are
+ * applied one after another i.e first the snapshot if it matches, the manifest
+ * file matcher and finally the manifest matcher. The matchers are related with
+ * AND relationship i.e. if snapshot matcher is set the snapshot containing a
+ * manifest_file must match. Returned result is an outcome of depth first search
+ * traversal of metadata tree.
+ *
+ * Deduplication:
+ *
+ * Results are deduplicated i.e. if two or more snapshots include the same
+ * manifest file, only one of them will be included in the result, the same rule
+ * applies to the manifest. Deduplication is based on the path as it is a unique
+ * identifier of metadata entity.
+ *
+ * Ordering:
+ *
+ * The order in which results are collected matches the order of DFS metadata
+ * tree traversal with deduplication.
+ */
+template<result_type ResultT>
+struct metadata_query {
+    static constexpr result_type r_type{ResultT};
```

Comment:
Not suggesting a change, but it's interesting that this is really basically only used as a switch for execution. Wondering if it would work if we made the query non-templatized and passed the result_type explicitly to do_execute_query.. though i guess we also use this to check needs_manifest etc

Reply:
I was considering that, the most important part i use it for is to determine the result we need to return based on: 
```
template<result_type T>
struct results_container {};
```
this needs to be known at the compile time, this is why i decided to use templates here